src/ucs/datastruct/arbiter.c

```diff
@@ -230,19 +240,47 @@ void ucs_arbiter_group_desched_nonempty(ucs_arbiter_t *arbiter,
     ucs_arbiter_group_head_reset(head);
 }
 
+static inline void
+ucs_arbiter_remove_and_reset_if_scheduled(ucs_arbiter_elem_t *elem)
+{
+    if (ucs_unlikely(ucs_arbiter_group_head_is_scheduled(elem))) {
+         /* removing scheduled head of empty list */
+         ucs_list_del(&elem->list);
+         ucs_arbiter_group_head_reset(elem);
+     }
+}
+
+static inline void
+ucs_arbiter_group_head_replace(ucs_arbiter_group_t *group,
+                               ucs_arbiter_elem_t *group_head,
+                               ucs_arbiter_elem_t *new_group_head)
+{
+    /* check if this is really the group head */
+    ucs_assert(group->tail->next == group_head);
```

Comment:
what about making sure that the group isn't empty?
so, the assert can be replaced with:
```
ucs_assert(!ucs_arbiter_group_is_empty(group) &&
           ucs_arbiter_elem_is_only(group, group_head));
```

Reply:
ucs_arbiter_elem_is_only() is not exactly the same check, so added assertion that group is not empty